### ISSUE TITLE
Assorted issues/thoughts/suggestions

### ISSUE THREAD

**Issue by iandoug** [FontManager/master#39]
Don't want to create a whole string of small threads...

Anyway:

1. The Font Viewer applet has an Install button, but not the main app. It's more needed there :-)

2. Not sure why both "browse" and "manage" panes ... possibly they could be combined?

3. In Browse, I click font expecting to see the views from Manage....:-) (hence (2) )

4. I also find myself right-clicking on things expecting an action dialogue to pop up (eg install, open containing folder, whatever...)

5. Font Manager is slow on startup, I assume this is to do with the number of fonts I have (assorted collections gathered over 25 years). Don't know if it is rescanning database/filesystem or what ... possibly this could be made a manual action (Rescan now) rather than automatically on startup. There is a progress indicator sometimes which sometimes spins and sometimes not, but no numerical indicator (And I'm sure I saw some something in the source code for that today... dunno why not showing).

6. Sometimes Font Manager displays 100% black screen. And is reluctant to close. Possibly some GTK+ / KDE compatibility issue?

7. Fonts which do not have Latin basic text (eg fonts using Private Use area) or fonts that stick other stuff on Latin basic page obviously do not show name properly. So maybe name needs to be printed in plain text by default. Also on the font preview screens ... sometimes FM is slow to switch font previews and then I'm not sure which one I'm actually looking at, so printing the font name in plain System font in the preview will help.

8. Not sure if it handles different fonts with same name nicely...  eg Metropolis
https://creativemarket.com/MehmetRehaTugcu/270838-Metropolis-Typeface
vs
https://github.com/chrismsimpson/Metropolis
Or copies of same font in different folders.

Think that's all for now :-)

Cheers, Ian

**Comment by JerryCasiano** [comment]
1.) It's already there. Hit the add button. And it's not even needed since dragging and dropping files into the list will install them.

2 & 3) They can and probably will be in the future.

4.) Right click menus will come when they're needed, in other words, when features are added.

5.) Since you didn't state how many files you have or how they're stored. It's impossible to know, there are many things that could affect that.
The application isn't really intended for users with tens of thousands of fonts and will probably never be suitable for such a use case. In my testing, on a six year old Thinkpad, it takes the application around twenty seconds to fully load with 2500 fonts. So good enough for average users.

6.) Possibly. 

7.) That will likely change.

8.) Depends on metadata. If it's identical, you'll have issues and not just with Font Manager.

In any case, yes, the application is far, far from 1.0 ;-)

**Comment by iandoug** [comment]
re (1), if you mean the "+" button, that brings up a "find file" dialogue, which is the problem ... I don't  where the font is or what the file name is. That's why I look for the "install" button... :-)

If you mean drag-and-drop from for example the Browse view, onto the System or Personal collections... that's something I never thought of doing... will try.

Currently have over 20k font files... so need to find a way to sort them sanely. Most are not installed (less 1000, and a lot of those were by mistake when I clicked the wrong thing while learning your program... :-) )

Thanks, Ian

**Comment by JerryCasiano** [comment]
No I mean you can drag files, folders or archives from a file manager onto the font list to install.

Yeah, 20k+ fonts is not something we'll ever handle well. It's not a goal.

The application uses Pango and Fontconfig same as the rest of a GTK+ environment, this is easier than doing our own rendering and means that WYSIWYG, if a font doesn't render properly it won't render properly on your desktop either. But it does mean that large amounts of fonts like that will cause slowness.

If we did our own rendering and went strictly off on our own database we could avoid that overhead and handle a large amount of fonts without slowdown. But as I said it's not a goal.

The application really is intended for your average desktop user and they typically don't have large quantities of fonts like you do, matter of fact most professionals probably don't have that amount of fonts.

**Comment by iandoug** [comment]
I have a lot of fonts dating back to my DOS days. Also things like a CD ROM with 4000 fonts, that I need to sort to find the good stuff and dump the junk. Etc. Not that I use all these fonts a lot, but there are times when I or my kid needs a particular kind of font then we need to way to search what we have for it...
FWIW I've decided to try and build a database of fonts (in general, as opposed to "what I have") and see if it is possible to find a way to automagically/semi-automagically categorise them. A process made more difficult by the lack of a standard way of categorising fonts ... Even fonts.com came up with their own way of doing it. So am pondering a structure for that first ... and then will need a way of programmatically checking glyphs to see what they have... which may not be possible without a lot of AI.

**Comment by JerryCasiano** [comment]
I think you'll find that the main issue with categorizing font files automatically is all the different file types and the varying level of quality.

Checking glyphs on a Freetype or Fontconfig system is pretty straightforward, so that shouldn't present much of an issue. What to do with that information is not so straightforward, aside from classifying based on orthography not sure there's any real use case there.

In any case, I'm currently reworking the lower layers. And to me, this is one the most difficult decisions. How much information to extract and keep. No sense spending time gathering and storing information we'll never actually use.

So if you come up with anything on that front please share it here.

Right now our current database is very simple you can open it and easily see what's there. It should give you an idea of the percentage of fonts which either don't have certain information or worse have the wrong information embedded in them.

**Comment by iandoug** [comment]
Not sure what you mean by "checking glyphs on a Freetype system" being pretty straightforward. What I was tentatively considering doing is automating the sort of measurements that PANOSE proposed.
[https://www.w3.org/Printing/stevahn.html](url)

But don't know enough about graphics processing / working with glyphs to know how easy/impossible that is. Idea was to classify by type of serif, eg 
[https://s-media-cache-ak0.pinimg.com/originals/8f/82/4d/8f824d07b6465b85ec88aab60fbc98ae.jpg](url)
plus a few more for sans-serif eg square, angled, concave, convex. 

But at the moment that's beyond what I know so I'm going to take a more iterative semi-automated approach, as follows:

1. rename all font files to have sensible names eg arialbi.ttf --> [Arial-BoldItalicMT.ttf](url) based on defined FontName (or possibly the Human version).
2. stick all these things in a database, including family name and whatever else in meta data seems reasonable. I have seen lots of bad info in the meta data myself so understand your reluctance to trust it :-)
3. build a web interface ('cos that's what I know, as opposed to native GUI) system to present a font, allow me to classify it manually into broad categories (eg sans, serif, blackletter, script, symbol, graphic), and then move font to suitably-named folder. Then check database for all other occurences  of same family and also move them. Delete duplicates based on having same hash.

Once that is done I can look at doing it again on each sorted folder, into finer categories. Thus iterative. What what I learn at each step helps for the next one.

I've got an API key from fonts.com which will allow me to interrogate their meta-data for zillions of commercial fonts, but this is further down the road.

Have been studying up on the various classification attempts. This paper was useful, but I disagree a bit with their final recommendation.
[http://luc.devroye.org/ParsonsJournalForInformationMapping-Childers+Griscti+Leben-2013.pdf](url)

**Comment by JerryCasiano** [comment]
I see. Something similar to what https://www.myfonts.com/WhatTheFont/ does.

Thanks for the links.

**Comment by iandoug** [comment]
I would image that those identify-the-font things work by isolating the glyphs then running some sort of fuzzy pattern matching against their known database of fonts.
I'm more interested in a different angle along the lines of "show me options for Serif:Didone" etc.
The font sites I've seen don't do that level of detail, and tend to be either commercial or freeware/indie rather than covering both universes.
Not planning on offering fonts for download, that's a whole different mine field :-)
